File: effs/words_fill_at.js
// Read exactly n bytes at fileOff into a word-aligned packed Array range.
function words_fill_at(path, fileOff, a, start, n) {
  const fs = require("fs");
  let fd;
  try {
    fd = fs.openSync(path, "r");
    const buf = Buffer.allocUnsafe(n);
    let done = 0;
    while (done < n) {
      const got = fs.readSync(fd, buf, done, n - done, fileOff + done);
      if (got <= 0) throw Object.assign(new Error("short read"), { errno: 5 });
      done += got;
    }
    for (let i = 0; i < n; i += 4) {
      a[(start + i) >> 2] = ((buf[i] | 0) | ((buf[i + 1] | 0) << 8) |
        ((buf[i + 2] | 0) << 16) | ((buf[i + 3] | 0) << 24)) >>> 0;
    }
    fs.closeSync(fd);
    return io_tup(a, io_done({ $: "Unit" }));
  } catch (e) {
    if (fd !== undefined) try { fs.closeSync(fd); } catch (_) {}
    return io_tup(a, io_fail(Math.abs(e.errno ?? 5)));
  }
}

File: effs/words_fill_at.c
// Read exactly n bytes at file_off directly into a word-aligned range of an
// existing packed Array. The Array comes back outside the Result.
Term words_fill_at_run(Env e, Term* f, IoWork* w) {
  uint64_t path_n = 0;
  char* path = io_cstr(e, f[0], &path_n);
  uint64_t file_off = (uint64_t)(u32)f[1];
  Term arr = f[2];
  uint64_t start = (uint64_t)(u32)f[3];
  uint64_t count = (uint64_t)(u32)f[4];
  int fd = io_nul(path, path_n) ? -1 : open(path, O_RDONLY);
  if (fd < 0) {
    int err = errno != 0 ? errno : ENOENT;
    free(path);
    return io_tup(e, arr, io_fail(e, (u32)err, NULL));
  }
  free(path);

  uint8_t* dst = (uint8_t*)blk_ptr(e.mem, term_loc(arr), 0) + start;
  uint64_t done = 0;
  while (done < count) {
    ssize_t got = pread(fd, dst + done, count - done, (off_t)(file_off + done));
    if (got < 0 && errno == EINTR) continue;
    if (got <= 0) {
      int err = got < 0 && errno != 0 ? errno : EIO;
      close(fd);
      return io_tup(e, arr, io_fail(e, (u32)err, NULL));
    }
    done += (uint64_t)got;
  }
  if (close(fd) != 0) {
    return io_tup(e, arr, io_fail(e, (u32)(errno != 0 ? errno : EIO), NULL));
  }
  return io_tup(e, arr, io_done(e, term_pak(CID_UNIT, 0)));
}

static void __attribute__((constructor)) words_fill_at_use(void) {
  io_eff(CID_WORDS_FILL_AT, words_fill_at_run, 0);
}
